EIPS/eip-5164.md

```diff
@@ -0,0 +1,211 @@
+---
+eip: 5164
+title: Cross-Chain Execution
+description: This specification defines an interface that supports execution across EVM networks.
+author: Brendan Asselstine (@asselstine), Anna Carroll (@anna-carroll), Hadrien Croubois (@Amxx), Nam Chu Hoai (@nambrot), Georgios (@geogons), Theo Gonella (@mintcloud), Rafael Solari (@rsolari), Auryn Macmillan (@auryn-macmillan), Nathan Ginnever (@nginnever)
+discussions-to: https://ethereum-magicians.org/t/eip-5164-cross-chain-execution/9658
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-06-14
+---
+
+## Abstract
+
+This specification defines a cross-chain execution interface for EVM-based blockchains. Developers will be able to have contracts call contracts on other chains. There are two parts to this spec: the "sending" interface and the "receiving" interface. The "sending" interface specifies the interface to send a call to another chain; this interface must be implemented by bridges. The "receiving" interface must be implemented by the developer; it is an interface for receiving cross-chain calls. This specification is agnostic of the bridge implementation; it simply standardizes the call interfaces.
+
+## Motivation
+
+Many Ethereum protocols need to coordinate state changes across multiple EVM-based blockchains. These chains often have native or third-party bridges that allow Ethereum contracts to execute code. However, bridges have different APIs so bridge integrations are bespoke.
+
+Bridge technology is improving rapidly, and there are multiple bridges to choose from. Each one affords different properties; with varying degrees of security, speed, and control.
+
+By standardizing a cross-chain execution interface, we can cleanly separate the transport layer from the application layer. Cross-chain execution becomes another composable piece with which the Ethereum ecosystem can build shared infrastructure.
+
+## Specification
+
+This specification allows contracts on one chain to send messages to contracts on another chain. There are two key contracts:
+
+- CrossChainRelayer
+- CrossChainReceiver
+
+The `CrossChainRelayer` lives on the origin chain. Users can share a single `CrossChainRelayer` or deploy their own.
+
+The `CrossChainReceiver` lives on the destination chain and receives relayed calls.
```

Comment:
`CrossChainRelayer` describes who is responsible for deploying it (users). Who is responsible for deploying `CrossChainReceiver`?

Reply:
The `CrossChainReceiver` interface has been renamed `CrossChainExecutor`.
To receive and execute messages on the receiving chain, users can deploy their own `CrossChainExecutor` or share a single one.

We have modified the specification of this interface and users need to emit the `ExecutedCalls` event once a Call has been executed but they don't need to implement a `receiveCalls` function to receive calls.
We have also added an optional `ExecutorAware` interface that would follow EIP-2771 and allow users to authenticate/verify where the call originated from.

This way, users are not constrained in the way they want to receive and execute calls.

For example, a user may use a pre-existing `CrossChainExecutor` contract to receive calls that would then be executed and call a function on an `ExecutorAware` contract. Using EIP-2771, they can authenticate the call and verify that it originated from the `CrossChainExecutor` contract they decided to use.
But it would also be possible for the user to only deploy their own `CrossChainExecutor` contract and this contract would receive calls and execute them but they wouldn't call an `ExecutorAware` contract and would instead call a function in the `CrossChainExecutor` contract.

Reply:
I've explained who needs to deploy a `CrossChainExecutor` contract in this commit: https://github.com/ethereum/EIPs/pull/5164/commits/772f1c7b03c55496f2f724073e8ebb227e154a5c#diff-41efe0617f861cc7330b025ecc86d55fe777d3c503208d83dca04258de4f5f41R36